### PR TITLE
session: fix bootstrap panic when upgrading to master

### LINKED ISSUE #56489: Invalid memory address or nil pointer when upgrade to master version
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
use test;
create database if not exists v1static;
set global tidb_analyze_version = 1;
set global tidb_partition_prune_mode = 'static';
set session tidb_enable_list_partition = 1;
create table rptcon (a bigint primary key, b varchar(20)) partition by range(a) (partition p0 values less than (10), partition p1 values less than (20), partition p2 values less than(30));
create table hptcon (a bigint primary key, b varchar(20)) partition by hash(a) partitions 3;
create table lptcon (a bigint primary key, b varchar(20)) partition by list(a) (partition p0 values in (1,2,3), partition p1 values in (14,16,18), partition p3 values in (20,25,27,29));
insert into rptcon values (1, 'a'), (2, 'b'), (3,'c'), (14, 'd'), (16, 'e'), (18, 'f'), (20, 'h'), (25, 'i'), (27, 'j'), (29, 'k');
insert into hptcon values (1, 'a'), (2, 'b'), (3,'c'), (14, 'd'), (16, 'e'), (18, 'f'), (20, 'h'), (25, 'i'), (27, 'j'), (29, 'k');
insert into lptcon values (1, 'a'), (2, 'b'), (3,'c'), (14, 'd'), (16, 'e'), (18, 'f'), (20, 'h'), (25, 'i'), (27, 'j'), (29, 'k');
drop stats rptcon;
drop stats hptcon;
drop stats lptcon;
analyze table rptcon  with 10 buckets;
analyze table hptcon  with 10 buckets;
analyze table lptcon  with 10 buckets;
```
before upgrade tidb version
```
Release Version: v8.0.0
Edition: Community
Git Commit Hash: 8ba1fa452b1ccdbfb85879ea94b9254aabba2916
Git Branch: HEAD
UTC Build Time: 2024-03-28 14:22:15
GoVersion: go1.21.4
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

### 2. What did you expect to see? (Required)
use operator upgrade cluster to master version success

### 3. What did you see instead (Required)
```
start tidb-server ...
/tidb-server --store=tikv --advertise-address=cluster-tidb-1.cluster-tidb-peer.oltp-global-stats-update-tps-7632425-1-763.svc --host=0.0.0.0 --path=cluster-pd:2379 --config=/etc/tidb/tidb.toml

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x3bb741a]

goroutine 1 [running]:
github.com/pingcap/tidb/pkg/infoschema.(*infoschemaV2).TableByID(0xc002177d48, {0x0, 0x0}, 0x74)
	/workspace/source/tidb/pkg/infoschema/infoschema_v2.go:625 +0x19a
github.com/pingcap/tidb/pkg/session.rebuildAllPartitionValueMapAndSorted(0xc00164c788)
	/workspace/source/tidb/pkg/session/bootstrap.go:3578 +0x193
github.com/pingcap/tidb/pkg/session.bootstrapSessionImpl({0x6fec9e0, 0xc001057320}, 0x689c448)
	/workspace/source/tidb/pkg/session/session.go:3508 +0x5f9
github.com/pingcap/tidb/pkg/session.BootstrapSession(...)
	/workspace/source/tidb/pkg/session/session.go:3397
main.createStoreAndDomain({0x0?, 0x675fc28?})
	/workspace/source/tidb/cmd/tidb-server/main.go:414 +0x1b4
main.main()
	/workspace/source/tidb/cmd/tidb-server/main.go:320 +0x3c5
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #56489 

Problem Summary:

### What changed and how does it work?

ctx is nil and it panic here:

https://github.com/pingcap/tidb/blob/1eb0c8c1235c21ebc0177db8e3deb4b05ef5bcba/pkg/infoschema/infoschema_v2.go#L625-L627

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Follow the upgrade steps described in the issue

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/session/bootstrap.go [hunk 1]
```diff
@@ -3575,7 +3575,11 @@ func rebuildAllPartitionValueMapAndSorted(s *session) {
 			if pi == nil || pi.Type != model.PartitionTypeList {
 				continue
 			}
-			tbl, ok := is.TableByID(s.currentCtx, t.ID)
+			ctx := s.currentCtx
+			if ctx == nil {
```
Comment (lance6716) [inline pkg/session/bootstrap.go:3579]: Can we change the caller to use a non-nil ctx? Because generally I assume the context should always be non-nil.

Comment (tiancaiamao) [reply pkg/session/bootstrap.go:3579]: I find this comment https://github.com/pingcap/tidb/blob/9b1fa1cda02d65f46f85b7f3c5b83d2bf96968bf/pkg/session/session.go#L180

Comment (tiancaiamao) [reply pkg/session/bootstrap.go:3579]: Maybe we just use context.Background() here? @lance6716

Comment (lance6716) [reply pkg/session/bootstrap.go:3579]: yes I think we can delete this comment and set context.Background() as the default value.

**Review by lance6716 — APPROVED** [review]

File: pkg/session/bootstrap.go [hunk 2]
```diff
@@ -3575,7 +3575,7 @@ func rebuildAllPartitionValueMapAndSorted(s *session) {
 			if pi == nil || pi.Type != model.PartitionTypeList {
 				continue
 			}
-			tbl, ok := is.TableByID(s.currentCtx, t.ID)
+			tbl, ok := is.TableByID(context.Background(), t.ID)
```
Comment (D3Hunter) [inline pkg/session/bootstrap.go:3578]: please pass `ctx` inside bootstrapSessionImpl to here

Comment (tiancaiamao) [reply pkg/session/bootstrap.go]: Done

**Review by D3Hunter — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
LGTM

**Pull request merged**